File: python/model/resonance_attention.py
import tensorflow as tf
import math
# Import using absolute path from rkm package root
from python.layers.scale_layer import ScaleLayer

def resonance_attention(config, inputs, mask, block_index, layers, training=False):
    """
    Multi-Head Resonance Attention mechanism using pre-instantiated Keras layers.
    Implements Resonance Attention Dynamics (Section 4 of formalism) using standard tf.matmul.
    Entropy calculation is approximated as per the original implementation.

    Args:
        config (dict): Configuration object containing 'embedding_dim', 'num_heads',
                       'sequence_length', 'beta' (though beta is expected in layer).
        inputs (tf.Tensor): Input hidden states, shape [batch, seq_len, embedding_dim].
        mask (tf.Tensor | None): Attention mask, shape [batch, 1, 1, seq_len] or similar broadcastable.
        block_index (int): Index of the block (potentially used for layer naming/selection).
        layers (dict): A dictionary containing pre-instantiated Keras layer instances.
                       Expected keys: 'query_projection', 'key_projection', 'value_projection',
                       'q_reshape', 'q_permute', 'k_reshape', 'k_permute', 'v_reshape', 'v_permute',
                       'scale_scores_layer', 'beta_sharpening_layer', 'softmax_layer',
                       'entropy_square_multiply', 'entropy_pool',
                       'output_permute', 'output_reshape', 'output_dense'.
        training (bool): Whether in training mode (currently unused in this function).

    Returns:
        dict: A dictionary containing:
              'output': The attention output tensor, shape [batch, seq_len, embedding_dim].
              'entropy': The entropy proxy tensor, shape [batch, num_heads].
              'attention_weights': The attention weights tensor, shape [batch, num_heads, seq_len, seq_len].
    """
    num_heads = config.get('num_heads')
    embedding_dim = config.get('embedding_dim')
    sequence_length = config.get('sequence_length') # May not be strictly needed if shapes are dynamic

    if not num_heads or not embedding_dim:
        raise ValueError("Missing 'num_heads' or 'embedding_dim' in config")

    head_dim = embedding_dim // num_heads
    if head_dim * num_heads != embedding_dim:
        raise ValueError("embedding_dim must be divisible by num_heads")

    # --- Check for required layers ---
    required_layers = [
        'query_projection', 'key_projection', 'value_projection',
        'q_reshape', 'q_permute', 'k_reshape', 'k_permute', 'v_reshape', 'v_permute',
        'scale_scores_layer', 'beta_sharpening_layer', 'softmax_layer',
        'entropy_square_multiply', 'entropy_pool',
        'output_permute', 'output_reshape', 'output_dense'
    ]
    for key in required_layers:
        if key not in layers:
            raise ValueError(f"Missing required layer '{key}' in layers dictionary")

    # --- Project Q, K, V ---
    query = layers['query_projection'](inputs) # Shape: [batch, seq, embed]
    key = layers['key_projection'](inputs)     # Shape: [batch, seq, embed]
    value = layers['value_projection'](inputs)   # Shape: [batch, seq, embed]

    # --- Reshape and Permute for Multi-Head ---
    # Reshape: [batch, seq, embed] -> [batch, seq, heads, head_dim]
    # Permute: [batch, seq, heads, head_dim] -> [batch, heads, seq, head_dim]
    def reshape_and_permute(tensor, reshape_layer, permute_layer):
        reshaped = reshape_layer(tensor)
        return permute_layer(reshaped)

    q = reshape_and_permute(query, layers['q_reshape'], layers['q_permute'])
    k = reshape_and_permute(key, layers['k_reshape'], layers['k_permute'])
    v = reshape_and_permute(value, layers['v_reshape'], layers['v_permute'])

    # --- Calculate Scaled Dot-Product Attention Scores ---
    # MatMul: q * k^T -> [batch, heads, seq, seq]
    # Use tf.matmul for standard attention calculation
    scores = tf.matmul(q, k, transpose_b=True)

    # Scale scores by 1 / sqrt(head_dim)
    # Assumes scale_scores_layer is initialized with scale_factor = 1.0 / math.sqrt(head_dim)
    scaled_scores = layers['scale_scores_layer'](scores)

    # Apply beta sharpening
    # Assumes beta_sharpening_layer is a ScaleLayer initialized with the desired beta factor
    sharpened_scores = layers['beta_sharpening_layer'](scaled_scores)

    # Apply mask if provided
    if mask is not None:
        # Ensure mask is broadcastable to scores shape [batch, heads, seq, seq]
        # Common mask shape: [batch, 1, 1, seq_len]
        # Add a large negative number where mask is zero (or False)
        sharpened_scores += (tf.cast(mask, dtype=sharpened_scores.dtype) * -1e9)

    # Apply softmax
    attention_weights = layers['softmax_layer'](sharpened_scores) # Shape: [batch, heads, seq, seq]

    # --- Calculate Entropy Proxy ---
    # H_proxy = mean(α^2) over seq, seq dimensions
    squared_weights = layers['entropy_square_multiply']([attention_weights, attention_weights])
    # Pool across the last two dimensions (seq, seq) -> [batch, heads]
    entropy = layers['entropy_pool'](squared_weights)

    # --- Calculate Weighted Values ---
    # MatMul: attention_weights * v -> [batch, heads, seq, head_dim]
    weighted_values = tf.matmul(attention_weights, v)

    # --- Reshape and Project Output ---
    # Permute: [batch, heads, seq, head_dim] -> [batch, seq, heads, head_dim]
    permuted_output = layers['output_permute'](weighted_values)
    # Reshape: [batch, seq, heads, head_dim] -> [batch, seq, embedding_dim]
    reshaped_output = layers['output_reshape'](permuted_output)
    # Final dense projection
    output = layers['output_dense'](reshaped_output) # Shape: [batch, seq, embedding_dim]

    return {
        'output': output,
        'entropy': entropy,
        'attention_weights': attention_weights
    }

# Example Layer Instantiation (Illustrative - specific layer configs depend on ResonantKnowledgeModel)
# batch_size = 4
# sequence_length = 128
# embedding_dim = 256
# num_heads = 8
# head_dim = embedding_dim // num_heads
# beta_value = 1.5 # Example beta
#
# layers_dict = {
#     'query_projection': tf.keras.layers.Dense(embedding_dim),
#     'key_projection': tf.keras.layers.Dense(embedding_dim),
#     'value_projection': tf.keras.layers.Dense(embedding_dim),
#     'q_reshape': tf.keras.layers.Reshape((sequence_length, num_heads, head_dim)),
#     'q_permute': tf.keras.layers.Permute((2, 1, 3)), # batch, heads, seq, head_dim
#     'k_reshape': tf.keras.layers.Reshape((sequence_length, num_heads, head_dim)),
#     'k_permute': tf.keras.layers.Permute((2, 1, 3)), # batch, heads, seq, head_dim
#     'v_reshape': tf.keras.layers.Reshape((sequence_length, num_heads, head_dim)),
#     'v_permute': tf.keras.layers.Permute((2, 1, 3)), # batch, heads, seq, head_dim
#     'scale_scores_layer': ScaleLayer(scale_factor=1.0 / math.sqrt(float(head_dim))),
#     'beta_sharpening_layer': ScaleLayer(scale_factor=beta_value),
#     'softmax_layer': tf.keras.layers.Softmax(axis=-1),
#     'entropy_square_multiply': tf.keras.layers.Multiply(),
#     'entropy_pool': tf.keras.layers.GlobalAveragePooling2D(data_format='channels_first'), # Pool over last 2 dims
#     'output_permute': tf.keras.layers.Permute((2, 1, 3)), # batch, seq, heads, head_dim
#     'output_reshape': tf.keras.layers.Reshape((sequence_length, embedding_dim)),
#     'output_dense': tf.keras.layers.Dense(embedding_dim)
# }

File: js/model/resonanceAttention.js
const tf = require('@tensorflow/tfjs-node');
const ScaleLayer = require('../layers/ScaleLayer'); // Need to import ScaleLayer

/**
 * Multi-Head Resonance Attention mechanism (Extracted from ResonantKnowledgeModel)
 * Implements Resonance Attention Dynamics (Section 4 of formalism)
 * Note: Iterative refinement and dynamic beta (β⁽ᵗ⁾) are not feasible in symbolic mode.
 * Entropy calculation is approximated.
 * Uses dense layer simulation for QK^T and Weights*V due to symbolic tensor limitations.
 *
 * @param {Object} config - Configuration object (embeddingDim, numHeads, sequenceLength, beta).
 * @param {Tensor} inputs - Input hidden states [batch, seq_len, embeddingDim].
 * @param {Tensor|null} mask - Attention mask (currently unused).
 * @param {number} blockIndex - Index of the block (used for selecting layers if stored per block).
 * @param {Object} layers - Object containing pre-instantiated layer instances for this attention block.
 * @param {boolean} training - Whether in training mode (currently unused).
 * @returns {Object} - Object containing { output, entropy, attentionWeights }.
 */
function resonanceAttention(config, inputs, mask, blockIndex, layers, training = false) { // Added layers parameter
  const { numHeads, embeddingDim, sequenceLength } = config;
  const headDim = Math.floor(embeddingDim / numHeads);
  // Beta and namePrefix are handled by the layer construction in the main class

  // Beta sharpness factor is now part of the pre-instantiated betaSharpeningLayer
  // const beta = config.beta || 1.0;
  // Name prefix is handled during layer construction
  // const namePrefix = `block_${blockIndex}_attn`;

  // Create Q, K, V projections
  // Use pre-instantiated layers
  const query = layers.queryProjection.apply(inputs);

  const key = layers.keyProjection.apply(inputs);

  const value = layers.valueProjection.apply(inputs);

  // Reshape for multi-head attention
  // Helper function inside to avoid polluting scope, or define outside if preferred
  // Reshaping and permuting is now done using pre-instantiated layers
  const reshapeAndPermute = (tensor, reshapeLayer, permuteLayer) => {
      const reshaped = reshapeLayer.apply(tensor);
      return permuteLayer.apply(reshaped);
  };


  // Apply reshape and permute with unique names
  // Apply reshape and permute using corresponding layers
  const q = reshapeAndPermute(query, layers.qReshape, layers.qPermute);
  const k = reshapeAndPermute(key, layers.kReshape, layers.kPermute);
  const v = reshapeAndPermute(value, layers.vReshape, layers.vPermute);


  // Calculate Attention Scores: softmax(β * QK^T / sqrt(headDim)) * V
  // Using dense layer simulation for QK^T due to symbolic tensor limitations

  // Reshape Q to 3D: [batch*heads, seq_len, head_dim]
  // Input q is [batch, heads, seq, headDim]
  // Target shape needs to combine batch and heads dimension.
  // tf.reshape can use -1 to infer dimensions.
  // Let's try reshaping q to [batch * heads, sequenceLength, headDim]
  const reshapedQ = layers.qReshape3D.apply(q);

  // Simulate QK^T using a dense layer projecting headDim -> seqLength
  // Input: [batch*heads, seq_len, headDim] -> Output: [batch*heads, seq_len, seqLength]
  const qkSimulated = layers.qkDenseSim.apply(reshapedQ);

  // Reshape scores back to 4D: [batch, heads, seq_len, seq_len]
  const dotProduct = layers.scoresReshape4D.apply(qkSimulated);

  // Scale scores
  // Scale factor is part of the pre-instantiated layer
  const scaledDotProduct = layers.scaleScoresLayer.apply(dotProduct);

  // Apply beta sharpening (using the fixed beta)
  // Beta factor is part of the pre-instantiated layer
  const sharpenedScores = layers.betaSharpeningLayer.apply(scaledDotProduct);

  // Apply mask if provided (TODO: Implement mask application)
  // let maskedScores = sharpenedScores;
  // if (mask) { ... }

  // Apply softmax to get attention weights
  const attentionWeights = layers.softmaxLayer.apply(sharpenedScores);

  // Calculate entropy using a layer
  // Entropy Calculation (Approximation for H = -∑ α log(α))
  // Using mean squared value of weights as a proxy (lower for sharper distributions)
  const squaredWeights = layers.entropySquareMultiply.apply([attentionWeights, attentionWeights]);
  // Pool across the last two dimensions (seq, seq).
  // Input shape: [batch, heads, seq, seq]. Output: [batch, heads]
  const entropyProxy = layers.entropyPool.apply(squaredWeights);
  // Maybe average across heads too? Let's keep it [batch, heads] for now.
  const entropy = entropyProxy;

  // Apply attention weights to values: Attn(x) = Weights * V
  // Using dense layer simulation for Weights * V

  // Reshape V to 3D: [batch*heads, seq_len, head_dim]
  const reshapedV = layers.vReshape3D.apply(v);
  // Reshape weights to 3D: [batch*heads, seq_len, seqLength]
  const reshapedWeights = layers.weightsReshape3D.apply(attentionWeights);

  // Simulate Weights * V using a dense layer projecting seqLength -> headDim
  // Input: [batch*heads, seq_len, seqLength] -> Output: [batch*heads, seq_len, headDim]
  const weightedValues3D = layers.weightsVDenseSim.apply(reshapedWeights);

  // Reshape back to 4D: [batch, heads, seq_len, head_dim]
  const weightedValues = layers.weightedValsReshape4D.apply(weightedValues3D);

  // Concatenate heads: [batch, heads, seq_len, head_dim] -> [batch, seq_len, heads, head_dim] -> [batch, seq_len, embedding_dim]
  // Permute dims: [batch, seq_len, heads, head_dim]
  const permutedOutput = layers.outputPermute.apply(weightedValues);

  // Reshape to [batch, seq_len, embedding_dim]
  const reshapedOutput = layers.outputReshape.apply(permutedOutput);

  // Final dense projection for the attention block output
  const output = layers.outputDense.apply(reshapedOutput);

  // Note: During training we'd update beta in a callback

  return { output, entropy, attentionWeights };
}

module.exports = resonanceAttention;